[WIP] Get started docs (#14807)
huggingface/transformers: docs/source/quicktour.mdx

[[open-in-colab]]

Get up and running with 🤗 Transformers! Start using the [`pipeline`] for rapid inference, and quickly load a pretrained model and tokenizer with [AutoClasses](./model_doc/auto) to solve your text, vision or audio task.

<Tip>
All code examples presented in the documentation have a toggle on the top left for PyTorch and TensorFlow. If
not, the code is expected to work for both backends without any change.
</Tip>

> # Pipeline

Reviewer: ```suggestion
## Pipeline
```

**Audio**:
* Audio classification: assign a label to a given segment of audio.
* Automatic speech recognition (ASR): transcribe audio data into text.

<Tip>
For more details about the [`pipeline`] and associated tasks, refer to the documentation [here](./main_classes/pipelines)
</Tip>

## Using the pipeline

In the following example, we will look at how to use the [`pipeline`] for sentiment analysis.

Install the following dependencies if you haven't already:

```bash
pip install torch
===PT-TF-SPLIT===
pip install tensorflow
```

Import the [`pipeline`] with the task you want to complete:

```python
>>> from transformers import pipeline
>>> classifier = pipeline("sentiment-analysis")
```

The pipeline downloads and caches a default [pretrained model](https://huggingface.co/distilbert-base-uncased-finetuned-sst-2-english) and tokenizer for sentiment analysis. Now you can use the `classifier` on your target text:

```python
>>> classifier("We are very happy to show you the 🤗 Transformers library."")
[{"label": "POSITIVE", "score": 0.9998}]
```

For more than one sentence, pass a list of sentences to the [`pipeline`] which returns a list of dictionaries:

> ```python

Reviewer: py is perfectly valid here, there is no reason to switch to python.

>>> results = classifier(["We are very happy to show you the 🤗 Transformers library.",
...            "We hope you don't hate it."])
>>> for result in results:
...     print(f"label: {result["label"]}, with score: {round(result["score"], 4)}")
label: POSITIVE, with score: 0.9998
label: NEGATIVE, with score: 0.5309
```

The [`pipeline`] can also iterate over an entire dataset. Start by installing the 🤗 Datasets library:

```bash
pip install datasets 
```

Create a [`pipeline`] with the task you want to solve for, and the model you want to use:

```python

> >>> from transformers import pipeline

Reviewer: We generally style doc samples to have one empty line between import and code (like black would do).

> >>> pipe = pipeline("automatic-speech-recognition", model="facebook/wav2vec2-base-960h", device=0)

Reviewer: Two things:
- `device=0` has not been explained yet
- Let's give this a descriptive name and not `pipe`. `speech-recognizer` for instance.

```

> Load the dataset you want to iterate over:

Reviewer: The documentation to `load_dataset` or the quicktour of Datasets should be linked here.

```python
>>> import datasets
>>> dataset = datasets.load_dataset("superb", name="asr", split="test")
```

> Now iterate over the dataset with the pipeline. This should be just as fast as running a custom loop on a GPU.

Reviewer: GPU support has not been introduced at this stage, and a random user running this code probably won't have one.

```python
# KeyDataset (only _pt_) will simply return the item in the dict returned by the dataset item

> # as we're not interested in the _target_ part of the dataset.

Reviewer: Let's not begin a code sample with a comment, we can put the same thing in prose before.

>>> for out in tqdm.tqdm(pipe(KeyDataset(dataset, "file"))):
...    print(out)
...    # {"text": "HE HOPED THERE WOULD BE STEW FOR DINNER TURNIPS AND CARROTS AND BRUISED POTATOES AND FAT MUTTON PIECES TO BE LADLED OUT IN THICK PEPPERED FLOWER FAT AND SAUCE"}
...    # {"text": ....}

> ...    # ....

Reviewer: If those are the outputs there should be not `...` in front of them.

```python
>>> from transformers import AutoTokenizer, AutoModelForSequenceClassification
>>> model = AutoModelForSequenceClassification.from_pretrained(model_name)
>>> tokenizer = AutoTokenizer.from_pretrained(model_name)
===PT-TF-SPLIT===
>>> from transformers import AutoTokenizer, TFAutoModelForSequenceClassification
>>> model = TFAutoModelForSequenceClassification.from_pretrained(model_name, from_pt=True)
>>> tokenizer = AutoTokenizer.from_pretrained(model_name)
```

Then you can specify the model and tokenizer, and apply the `classifier` on your target text:

```python
>>> classifier = pipeline('sentiment-analysis', model=model, tokenizer=tokenizer)
>>> classifier("Nous sommes très heureux de vous présenter la bibliothèque 🤗 Transformers.")
[{"label": "5 stars", "score": 0.7272651791572571}]
```

If you can't find a model for your use-case, you will need to fine-tune a pretrained model on your data. Take a look at our [fine-tuning tutorial](./training) to learn how. Finally, after you've fine-tuned your pretrained model, please consider sharing it (see tutorial [here](./model_sharing)) with the community on the Model Hub to democratize NLP for everyone! 🤗

> # AutoClasses

Reviewer: Header level 2 at least.

<Youtube id="AhChOFRegn4"/>

Under the hood, the [`AutoTokenizer`] and [`AutoModel`] classes work together to power the [`pipeline`]. The [AutoClasses](./model_doc/auto) are shortcuts that automatically retrieves the model architecture of a pretrained model from it's name or path. You only need to select the appropriate [`AutoModel`] for your task and it's associated tokenizer with [`AutoTokenizer`]. 

Let's return to our example and see how we can use the `AutoClasses` to replicate the results of the [`pipeline`].

> ## Using the AutoTokenizer

Reviewer: Should be one more than before, so 3 here.

A tokenizer is responsible for preprocessing text in a way that is understandable by the model. First, the tokenizer will split a given text into words called *tokens*. There are multiple rules that govern the tokenization process, including how to split a word and at what level (learn more [here](./tokenizer_summary)). The most important thing to remember though is you need to instantiate the tokenizer with the name of the model to ensure you're using the same tokenization rules a model was pretrained with.

Load a tokenizer with [`AutoTokenizer`]:

```python
>>> from transformers import AutoTokenizer, AutoModelForSequenceClassification
>>> model_name = "nlptown/bert-base-multilingual-uncased-sentiment"
>>> tokenizer = AutoTokenizer.from_pretrained(model_name)
===PT-TF-SPLIT===
>>> from transformers import AutoTokenizer, TFAutoModelForSequenceClassification
>>> model_name = "nlptown/bert-base-multilingual-uncased-sentiment"
>>> tokenizer = AutoTokenizer.from_pretrained(model_name)
```

Next, the tokenizer converts the tokens into numbers in order to construct a tensor as input to the model. This is known as the model's *vocabulary*.

Pass your text to the tokenizer:

```python

> >>> input = tokenizer("We are very happy to show you the 🤗 Transformers library.")

Reviewer: `input` is a reserved word in Python so we should avoid using it. How about `encoding`?

... )
```

Read the [preprocessing](./preprocessing) tutorial for more details about tokenization.

> ## Using the AutoModel

Reviewer: One more level

```python
>>> model_name = "nlptown/bert-base-multilingual-uncased-sentiment"
>>> pt_model = AutoModelForSequenceClassification.from_pretrained(model_name)
===PT-TF-SPLIT===
>>> model_name = "nlptown/bert-base-multilingual-uncased-sentiment"
>>> # This model only exists in PyTorch, so we use the _from_pt_ flag to import that model in TensorFlow.
>>> tf_model = TFAutoModelForSequenceClassification.from_pretrained(model_name, from_pt=True)
```

<Tip>
See the [task summary](./task_summary) for which [`AutoModel`] to use for which task.
</Tip>

Now you can pass your preprocessed batch of inputs directly to the model. If you are using a PyTorch model, unpack the dictionary by adding `**`. For TensorFlow models, pass the dictionary keys directly to the tensors:

```python
>>> pt_outputs = pt_model(**pt_batch)
===PT-TF-SPLIT===
>>> tf_outputs = tf_model(tf_batch)
```

> The model outputs the model's final activations in the `logits` attribute. Apply the softmax activation to the `logits` to retrieve the predictions:

Reviewer: ```suggestion
The model outputs the model's final activations in the `logits` attribute. Apply the softmax function to the `logits` to retrieve the probabilities:
```
Softmax is not an activation, and it returns probabilities. Predictions would be the classes with the maximum probabilities/logits.

```python
>>> from torch import nn
>>> pt_predictions = nn.functional.softmax(pt_outputs.logits, dim=-1)
>>> print(pt_predictions)
tensor([[2.2043e-04, 9.9978e-01],
        [5.3086e-01, 4.6914e-01]], grad_fn=<SoftmaxBackward>)
===PT-TF-SPLIT===
>>> import tensorflow as tf
>>> tf_predictions = tf.nn.softmax(tf_outputs.logits, axis=-1)
>>> print(tf_predictios)
tf.Tensor(
[[2.2043e-04 9.9978e-01]
 [5.3086e-01 4.6914e-01]], shape=(2, 2), dtype=float32)
```

<Tip>

> All 🤗 Transformers models (PyTorch or TensorFlow) return the model activations *before* the final activation
> function (like softmax) because the final activation function is often fused with the loss.

Reviewer: ```suggestion
All 🤗 Transformers models (PyTorch or TensorFlow) outputs the tensors *before* the final activation
function (like softmax) because the final activation function is often fused with the loss.
```

</Tip>

> Models are standard [torch.nn.Module](https://pytorch.org/docs/stable/nn.html#torch.nn.Module) or [tf.keras.Model](https://www.tensorflow.org/api_docs/python/tf/keras/Model) so you can use them in your usual training loop. However, to make things easier, 🤗 Transformers provides a [`Trainer`] class for PyTorch that adds functionality for distributed training, mixed precision, and more. For TensorFlow, you can use the `fit()` method from [Keras](https://keras.io/). Refer to the [training tutorial](./training) for more details.

Reviewer: ```suggestion
Models are standard [`torch.nn.Module`](https://pytorch.org/docs/stable/nn.html#torch.nn.Module) or [`tf.keras.Model`](https://www.tensorflow.org/api_docs/python/tf/keras/Model) so you can use them in your usual training loop. However, to make things easier, 🤗 Transformers provides a [`Trainer`] class for PyTorch that adds functionality for distributed training, mixed precision, and more. For TensorFlow, you can use the `fit()` method from [Keras](https://keras.io/). Refer to the [training tutorial](./training) for more details.
```

<Tip>
PyTorch model outputs are special dataclasses so their attributes are autocompleted in an IDE.
The model outputs also behave like a tuple or a dictionary (e.g., you can index with an integer, a slice or a string) in which
case the attributes that are `None` are ignored.
</Tip>

> ## Save your model

Reviewer: Probably need one level more.

> Once your model is fine-tuned, you can save it with its tokenizer using `save_pretrained`:

Reviewer: Link to `PreTrainedModel.save_pretrained` here.
```suggestion
Once your model is fine-tuned, you can save it with its tokenizer using [`~PreTrainedModel.save_pretrained`]:
```

```python
>>> pt_save_directory = "./pt_save_pretrained"
>>> tokenizer.save_pretrained(pt_save_directory)
>>> pt_model.save_pretrained(pt_save_directory)
===PT-TF-SPLIT===
>>> tf_save_directory = "./tf_save_pretrained"
>>> tokenizer.save_pretrained(tf_save_directory)
>>> tf_model.save_pretrained(tf_save_directory)
```

> When you are ready to use the model again, reload it with `from_pretrained`:

Reviewer: Same here, link to the doc.

```py
```python
>>> from transformers import AutoModel
>>> tokenizer = AutoTokenizer.from_pretrained(tf_save_directory)
>>> pt_model = AutoModel.from_pretrained(tf_save_directory, from_tf=True)
>>> pt_model = AutoModelForSequenceClassification.from_pretrained(tf_save_directory, from_tf=True)
===PT-TF-SPLIT===
>>> from transformers import TFAutoModel
>>> tokenizer = AutoTokenizer.from_pretrained(pt_save_directory)
>>> tf_model = TFAutoModel.from_pretrained(pt_save_directory, from_pt=True)
```

Lastly, you can also ask the model to return all hidden states and all attention weights if you need them:

```py
>>> pt_outputs = pt_model(**pt_batch, output_hidden_states=True, output_attentions=True)
>>> all_hidden_states  = pt_outputs.hidden_states 
>>> all_attentions = pt_outputs.attentions
===PT-TF-SPLIT===
>>> tf_outputs = tf_model(tf_batch, output_hidden_states=True, output_attentions=True)
>>> all_hidden_states =  tf_outputs.hidden_states
>>> all_attentions = tf_outputs.attentions
```

> ### Accessing the code

Reviewer: I don't think we should remove this from here if we are not adding it somewhere else.

Author: For sure, we will add this section in the guides as an alternative way to load a model.
